### PR TITLE
Refactor fill component to avoid relying on componentWillUpdate and use React Hooks instead

### PR DESCRIPTION
**Pull request by youknowriad**
Refs #11360 

I'm not familiar with the details of the slot/fill implementation especially the occurences thing so please make sure to test this extensively.

**Testing instructions**

 - Check the block toolbars and inspectors
 - Check the plugin slots (install dropit or similar...)

### CODE REVIEW

**Comment by aduth** [comment]
> I'm not familiar with the details of the slot/fill implementation especially the occurences thing

Some more context at https://github.com/WordPress/gutenberg/pull/3472#issuecomment-344955291 . I think the main idea is to force all fills to render again via `key` when another fill is removed.

I worry a bit that it might rely on very specific lifecycle timing (the _before_ render behavior of `componentWillUpdate` vs. the _after_ behavior of `componentDidUpdate` / `useEffect`). I'd wonder too if there's some way this could be eliminated / reimplemented.

**Comment by youknowriad** [comment]
@aduth do you know if there's a test covering this as it would help here.

**Comment by aduth** [comment]
@youknowriad There was a test case added in the relevant commit, yes. In that case, we may be able to judge by this passing.

https://github.com/WordPress/gutenberg/commit/73c3e3095023350ddbaa56eebc060e070b4c8fd1#diff-a7594baf81b70f7a01447aee3dedcb2bR100

**Comment by youknowriad** [comment]
Ok tests are failing which is good as it confirms the issue but the issue is I have no idea how to fix this right now. I'll probably rethink about it tomorrow.

**Comment by youknowriad** [comment]
I give up on this one, I was not able to make the tests pass. I feel like I'm not far from finding a solution though.

**Comment by youknowriad** [comment]
I think I found a fix for the reordering issue, let's see what the tests say.

File: packages/components/src/slot-fill/fill.js [hunk 1]
```diff
@@ -15,64 +15,56 @@ import { Consumer } from './context';
...
 
-class FillComponent extends Component {
-	constructor() {
-		super( ...arguments );
-		this.occurrence = ++occurrences;
-	}
+function FillComponent( { name, getSlot, children, registerFill, unregisterFill } ) {
+	// Random state used to rerender the component if needed, ideally we don't need this
+	const [ , updateRerenderState ] = useState( {} );
+	const rerender = () => updateRerenderState( {} );
```
Comment (gziolo) [inline packages/components/src/slot-fill/fill.js]: According to @gaearon

https://twitter.com/dan_abramov/status/1118301474876948480

you can do this:

```js
const [, forceUpdate] = useReducer(x => x + 1, 0)
```

**Review by gziolo — APPROVED** [review]
It's complex stuff. I don't fully understand all changes applied but based on Travis reaction I'm happy with the proposal :)

I strongly recommend another pair of eyes on this PR before you proceed.

File: packages/components/src/slot-fill/fill.js (hunk 1 above)
Comment (youknowriad) [reply packages/components/src/slot-fill/fill.js]: yes, not sure which one is best. I guess it doesn't matter that much

**Comment by veeramarni** [comment]
@youknowriad just checking whether it is ever merged to master as I don't find it in the master that uses react-hooks.

**Comment by youknowriad** [comment]
Yes, this was merged. you're maybe looking at the wrong master (your fork or something?)

**Comment by veeramarni** [comment]
Oh sorry, i see it now.

**Comment by gziolo** [comment]
This RR might introduce a regression which isn't covered with tests:

#15641 and https://github.com/WordPress/gutenberg/pull/15516#issuecomment-496720735

**Comment by aduth** [comment]
> This RR might introduce a regression which isn't covered with tests:
> 
> #15641 and [#15516 (comment)](https://github.com/WordPress/gutenberg/pull/15516#issuecomment-496720735)

Per https://github.com/WordPress/gutenberg/pull/15541#issuecomment-490979404, it is intended to be covered with tests, though if it is in-fact a regression, it would appear coverage could be improved.

**Pull request merged**